File: examples/05-enums-generics/generated.user.c

typedef struct Box_Int32 Box_Int32;

struct Box_Int32
{
    Int32 value;
};

simple Box_Int32* Box_Int32_new(Int32 value)
{
    Box_Int32* self = (Box_Int32*)kira_rc_alloc_with(sizeof(Box_Int32), null);
    self->value = value;
    return self;
}

typedef enum BuildStatus
{
    BUILD_STATUS_READY,
    BUILD_STATUS_RUNNING,
    BUILD_STATUS_DONE
} BuildStatus;

Int32 id_Int32(Int32 value);
Int32 main(Void);

Int32 id_Int32(Int32 value)
{
    return value;
}

/* module app:box */
/* module app:main */
/* use app:status */
/* use app:box */
Int32 main(Void)
{
    BuildStatus state = BUILD_STATUS_READY;
    Box_Int32* wrapped = Box_Int32_new(7);
    Int32 value = id_Int32(wrapped->value);
    if((state == BUILD_STATUS_READY))
    {
        print("%d\n", value);
    }
    kira_rc_release(wrapped);
    return 0;
}
/* module app:status */

File: examples/05-enums-generics/generated.user.js

// module "kira:math"
function clamp(value, lo, hi) {
    return Math.max(lo, Math.min(value, hi));
}

function lerp(a, b, t) {
    return (a + ((b - a) * t));
}

// module "app:box"
class Box {
    constructor(value) {
        this.value = value;
    }
}

function id(value) {
    return value;
}

// module "app:main"
// use "app:status"
// use "app:box"
function main() {
    const state = BuildStatus.READY;
    const wrapped = new Box(7);
    const value = id(wrapped.value);
    if (op_eq(state, BuildStatus.READY)) {
        kira_trace(value);
    }
}

// module "app:status"
const BuildStatus = Object.freeze({ READY: 0, RUNNING: 1, DONE: 2 });

main();
